src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +616,303 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
+		 *      node is distributed too.
+		 *
+		 * ii)  Inner join:
+		 *
+		 *      The join node is recurring only if both sides of the join are
+		 *      recurring.
+		 *
+		 *      Otherwise, if at least one side of the join is distributed, then
+		 *      this would ensure not returning recurring tuples from the workers;
+		 *      hence makes the join node "not recurring".
+		 */
+		JoinExpr *joinExpr = (JoinExpr *) joinNode;
+		bool leftNodeIsRecurring = IsJoinNodeRecurring(joinExpr->larg, query);
+		bool rightNodeIsRecurring = IsJoinNodeRecurring(joinExpr->rarg, query);
+
+		switch (joinExpr->jointype)
+		{
+			case JOIN_LEFT:
+			{
+				return leftNodeIsRecurring;
+			}
+
+			case JOIN_RIGHT:
+			{
+				return rightNodeIsRecurring;
+			}
+
+			case JOIN_FULL:
+			{
+				return leftNodeIsRecurring || rightNodeIsRecurring;
+			}
+
+			case JOIN_INNER:
+			{
+				return leftNodeIsRecurring && rightNodeIsRecurring;
+			}
+
+			default:
+			{
+				ereport(ERROR, (errmsg("unexpected join type")));
+			}
+		}
+	}
+	else if (IsA(joinNode, RangeTblRef))
+	{
+		RangeTblRef *rangeTableRef = (RangeTblRef *) joinNode;
+		int rangeTableIndex = rangeTableRef->rtindex;
+		List *rangeTableList = query->rtable;
+		RangeTblEntry *rangeTableEntry = rt_fetch(rangeTableIndex, rangeTableList);
+
+		/*
+		 * Inspect whether this side of the JOIN clause is recurring,
+		 * as in RelationInfoContainsOnlyRecurringTuples.
+		 */
+		return !FindNodeMatchingCheckFunctionInRangeTableList(list_make1(rangeTableEntry),
+															  IsDistributedTableRTE);
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("unexpected join node type")));
+	}
+}
+
+
+/*
+ * RecursivelyPlanDistributedJoinNode is a helper function for
+ * RecursivelyPlanRecurringTupleOuterJoins that recursively plans given
+ * distributed node that is known to be inner side of an outer join.
+ */
+static void
+RecursivelyPlanDistributedJoinNode(Node *distributedNode, Query *query,
+								   RecursivePlanningContext *recursivePlanningContext)
+{
+	if (IsA(distributedNode, JoinExpr))
+	{
+		/*
+		 * This, for example, means that RecursivelyPlanRecurringTupleOuterJoins
+		 * needs to plan inner side, i.e., "<distributed> INNER JOIN <distributed>",
+		 * of the following join:
+		 *   <recurring> LEFT JOIN (<distributed> INNER JOIN <distributed>)
+		 *
+		 * XXX: Ideally we should handle such a sub join tree by moving
+		 *      it into a subquery "as a whole" but this implies that we need to
+		 *      rebuild the rtable and re-point all the Vars to the new rtable
+		 *      indexes, so we've not implemented that yet.
+		 *
+		 *      Instead, we recursively plan all the distributed tables in that
+		 *      sub join tree. This is much more inefficient than the other
+		 *      approach (since we lose the opportunity to push-down the whole
+		 *      sub join tree into the workers) but is easier to implement.
+		 */
+
+		Node *larg = ((JoinExpr *) distributedNode)->larg;
+		if (!IsJoinNodeRecurring(larg, query))
```

Comment:
We should have tests with DEBUG1 such that we show we do not recursively plan recurring nodes:
```SQL
SET client_min_messages TO DEBUG1;
SELECT count(*) FROM test_r_1 LEFT JOIN (test_r_2 JOIN test_2 USING(key)) USING  (key);
```
test_r_N tables are reference tables. Also, please add similar to rarg